File: gems/decomplex/rust/src/decomplex/syntax/adapters/php.rs
use super::super::tree_sitter_adapter::{named_children, AssignmentTarget, CallTarget, Target};
use super::super::Language;
use super::base::LanguageProfile;
use crate::decomplex::ast::{node_text, normalize_text};
use tree_sitter::{Language as TreeSitterLanguage, Node};

pub(crate) struct PhpProfile;

impl LanguageProfile for PhpProfile {
    fn language(&self) -> Language {
        Language::Php
    }

    fn grammar(&self) -> TreeSitterLanguage {
        tree_sitter_php::LANGUAGE_PHP.into()
    }

    fn function_node_kinds(&self) -> &[&str] {
        &["function_definition", "method_declaration"]
    }

    fn class_owner_node_kinds(&self) -> &[&str] {
        &["class_declaration"]
    }

    fn parameter_list_node_kinds(&self) -> &[&str] {
        &["formal_parameters"]
    }

    fn parameter_identifier_node_kinds(&self) -> &[&str] {
        &["name", "variable_name"]
    }

    fn function_body_node_kinds(&self) -> &[&str] {
        &["compound_statement", "declaration_list"]
    }

    fn call_node_kinds(&self) -> &[&str] {
        &[
            "function_call_expression",
            "member_call_expression",
            "scoped_call_expression",
            "print_intrinsic",
        ]
    }

    fn identifier_node_kinds(&self) -> &[&str] {
        &["name", "variable_name"]
    }

    fn assignment_node_kinds(&self) -> &[&str] {
        &["assignment_expression", "augmented_assignment_expression"]
    }

    fn assignment_operator_tokens(&self) -> &[&str] {
        &["=", "+=", "-=", "*=", "/=", "%="]
    }

    fn path_action_node_kinds(&self) -> &[&str] {
        &[
            "function_call_expression",
            "member_call_expression",
            "scoped_call_expression",
            "expression_statement",
            "return_statement",
            "print_intrinsic",
        ]
    }

    fn simple_action_wrapper_node_kinds(&self) -> &[&str] {
        &["compound_statement", "declaration_list"]
    }

    fn comparison_node_kinds(&self) -> &[&str] {
        &["binary_expression"]
    }

    fn comparison_operators(&self) -> &[&str] {
        &["==", "!=", "===", "!==", "<", "<=", ">", ">="]
    }

    fn case_node_kinds(&self) -> &[&str] {
        &["switch_statement"]
    }

    fn case_arm_node_kinds(&self) -> &[&str] {
        &["case_statement"]
    }

    fn case_container_stop_node_kinds(&self) -> &[&str] {
        &[
            "function_definition",
            "method_declaration",
            "class_declaration",
        ]
    }

    fn case_subject_skip_node_kinds(&self) -> &[&str] {
        &["case_statement", "else", "comment"]
    }

    fn default_case_patterns(&self) -> &[&str] {
        &["_", "default"]
    }

    fn boolean_and_operators(&self) -> &[&str] {
        &["&&", "and"]
    }

    fn boolean_container_node_kinds(&self) -> &[&str] {
        &["binary_expression"]
    }

    fn parenthesized_wrapper_node_kinds(&self) -> &[&str] {
        &["parenthesized_expression"]
    }

    fn branch_node_kinds(&self) -> &[&str] {
        &["if_statement", "foreach_statement", "switch_statement"]
    }

    fn field_like_node_kinds(&self) -> &[&str] {
        &[
            "member_access_expression",
            "nullsafe_member_access_expression",
            "member_call_expression",
            "class_constant_access_expression",
        ]
    }

    fn normalize_source_text(&self, text: &str) -> String {
        normalize_text(&php_normalize_source(text))
    }

    fn normalize_parameter_name(&self, text: &str) -> String {
        php_identifier_text_value(text)
    }

    fn normalize_local_identifier_text(&self, text: &str) -> String {
        php_identifier_text_value(text)
    }

    fn function_name(&self, node: Node<'_>, source: &str) -> Option<String> {
        if self.function_node_kinds().contains(&node.kind()) {
            return node
                .child_by_field_name("name")
                .or_else(|| php_first_name_node(node))
                .and_then(|name| php_name_text(name, source));
        }
        self.default_function_name(node, source)
    }

    fn owner_name_from_declaration(&self, node: Node<'_>, source: &str) -> Option<String> {
        if node.kind() == "class_declaration" {
            return node
                .child_by_field_name("name")
                .or_else(|| php_first_name_node(node))
                .and_then(|name| php_name_text(name, source));
        }
        self.default_owner_name_from_declaration(node, source)
    }

    fn function_visibility(&self, node: Node<'_>, source: &str) -> Option<String> {
        named_children(node)
            .into_iter()
            .find(|child| child.kind() == "visibility_modifier")
            .map(|modifier| node_text(modifier, source).to_string())
            .filter(|modifier| matches!(modifier.as_str(), "public" | "private" | "protected"))
            .or_else(|| Some("public".to_string()))
    }

    fn assignment_target<'tree>(&self, node: Node<'tree>) -> Option<AssignmentTarget<'tree>> {
        self.default_assignment_target(node)
    }

    fn call_target<'tree>(&self, node: Node<'tree>, source: &str) -> Option<CallTarget<'tree>> {
        if !self.call_node_kinds().contains(&node.kind()) {
            return None;
        }
        if node.kind() == "print_intrinsic" {
            return Some(CallTarget::new(
                "self".to_string(),
                "print".to_string(),
                php_print_argument_texts(node, source),
            ));
        }
        let mut target = self.default_call_target(node, source)?;
        target.receiver = php_normalize_receiver(&target.receiver);
        target.message = php_identifier_text_value(&target.message);
        Some(target)
    }

    fn call_argument_texts(&self, node: Node<'_>, source: &str) -> Vec<String> {
        self.call_argument_nodes(node)
            .into_iter()
            .map(|argument| normalize_text(&php_normalize_source(node_text(argument, source))))
            .collect()
    }

    fn state_target(&self, lhs: Node<'_>, source: &str) -> Option<Target> {
        let target = self.default_state_target(lhs, source)?;
        Some(Target {
            receiver: php_normalize_receiver(&target.receiver),
            field: php_identifier_text_value(&target.field),
        })
    }

    fn state_read_target(&self, node: Node<'_>, source: &str) -> Option<Target> {
        let target = self.default_state_read_target(node, source)?;
        Some(Target {
            receiver: php_normalize_receiver(&target.receiver),
            field: php_identifier_text_value(&target.field),
        })
    }

    fn member_field_text(&self, field: Node<'_>, source: &str) -> Option<String> {
        php_name_text(field, source)
    }

    fn case_pattern_texts(&self, patterns: &[Node<'_>], source: &str) -> Vec<String> {
        patterns
            .iter()
            .map(|pattern| normalize_text(&php_normalize_source(node_text(*pattern, source))))
            .collect()
    }
}

fn php_first_name_node<'tree>(node: Node<'tree>) -> Option<Node<'tree>> {
    named_children(node)
        .into_iter()
        .find(|child| php_name_node(*child))
}

fn php_name_node(node: Node<'_>) -> bool {
    matches!(node.kind(), "name" | "qualified_name" | "variable_name")
}

fn php_name_text(node: Node<'_>, source: &str) -> Option<String> {
    let text = php_identifier_text_value(node_text(node, source));
    (!text.is_empty()).then_some(text)
}

fn php_identifier_text_value(text: &str) -> String {
    text.trim().trim_start_matches('$').to_string()
}

fn php_normalize_receiver(receiver: &str) -> String {
    let value = php_normalize_source(&php_identifier_text_value(receiver));
    if value == "this" {
        "self".to_string()
    } else {
        value
    }
}

fn php_print_argument_texts(node: Node<'_>, source: &str) -> Vec<String> {
    named_children(node)
        .into_iter()
        .flat_map(|child| {
            if child.kind() == "parenthesized_expression" {
                let children = named_children(child);
                if children.is_empty() {
                    vec![child]
                } else {
                    children
                }
            } else {
                vec![child]
            }
        })
        .map(|argument| normalize_text(&php_normalize_source(node_text(argument, source))))
        .collect()
}

fn php_normalize_source(source: &str) -> String {
    let mut out = String::new();
    let mut chars = source.chars().peekable();
    while let Some(ch) = chars.next() {
        if ch == '$' {
            if chars
                .peek()
                .map(|next| *next == '_' || next.is_ascii_alphabetic())
                .unwrap_or(false)
            {
                continue;
            }
        }
        out.push(ch);
    }
    out.replace("->", ".").replace("::", ".")
}

File: gems/decomplex/lib/decomplex/syntax/php.rb
# frozen_string_literal: true

module Decomplex
  module Syntax
    PHP_LEXICON = LanguageLexicon.new(
      nil_literal_patterns: [/\bnull\b/i].freeze,
      type_guard_patterns: [
        /\bnull\b/i,
        /\b(?:is_null|isset|empty|is_a|instanceof)\s*(?:\(|\b)/
      ].freeze,
      diagnostic_patterns: [
        /\bthrow\b/,
        /\b(?:die|exit|trigger_error)\s*\(/
      ].freeze,
      trivial_patterns: [
        /\A(?:null|true|false|0|1|break|continue)\s*;?\z/i,
        /\Areturn\s+(?:null|true|false|0|1)\s*;?\z/i
      ].freeze
    ).freeze

    class PhpSyntaxAdapter < TreeSitterLanguageAdapter
      FUNCTION_NODE_KINDS = %w[function_definition method_declaration].freeze
      CALL_NODE_KINDS = %w[function_call_expression member_call_expression scoped_call_expression print_intrinsic].freeze
      CLASS_OWNER_NODE_KINDS = %w[class_declaration].freeze
      PARAMETER_LIST_NODE_KINDS = %w[formal_parameters].freeze
      FUNCTION_BODY_NODE_KINDS = %w[compound_statement declaration_list].freeze
      IDENTIFIER_NODE_KINDS = %w[name variable_name].freeze
      FIELD_IDENTIFIER_NODE_KINDS = [].freeze
      PARAMETER_IDENTIFIER_NODE_KINDS = %w[name variable_name simple_parameter].freeze
      LOCAL_DECLARATION_NODE_KINDS = [].freeze
      LOCAL_VARIABLE_DECLARATOR_NODE_KINDS = [].freeze
      FIELD_DECLARATION_NODE_KINDS = %w[property_declaration].freeze
      DECLARATION_SITE_PARENT_NODE_KINDS = %w[simple_parameter method_declaration function_definition class_declaration].freeze
      ASSIGNMENT_NODE_KINDS = %w[assignment_expression augmented_assignment_expression].freeze
      ASSIGNMENT_STATE_DECLARATION_NODE_KINDS = %w[assignment_expression].freeze
      ASSIGNMENT_OPERATOR_TOKENS = %w[= += -= *= /= %=].freeze
      PATH_ACTION_NODE_KINDS = %w[function_call_expression member_call_expression scoped_call_expression expression_statement return_statement print_intrinsic].freeze
      SIMPLE_ACTION_WRAPPER_NODE_KINDS = %w[compound_statement declaration_list].freeze
      COMPARISON_NODE_KINDS = %w[binary_expression].freeze
      BRANCH_NODE_KINDS = %w[if_statement foreach_statement switch_statement].freeze
      LOOP_NODE_KINDS = %w[foreach_statement].freeze
      BRANCH_LOOP_NODE_KINDS = LOOP_NODE_KINDS
      CASE_NODE_KINDS = %w[switch_statement].freeze
      BRANCH_CASE_NODE_KINDS = %w[switch_statement].freeze
      IF_NODE_KINDS = %w[if_statement].freeze
      CASE_ARM_NODE_KINDS = %w[case_statement].freeze
      SWITCH_CASE_ARM_NODE_KINDS = %w[case_statement].freeze
      CASE_CONTAINER_STOP_NODE_KINDS = %w[function_definition method_declaration class_declaration].freeze
      CASE_SUBJECT_SKIP_NODE_KINDS = %w[case_statement else comment].freeze
      DEFAULT_CASE_PATTERNS = %w[_ default].freeze
      BOOLEAN_AND_OPERATORS = %w[&& and].freeze
      BOOLEAN_CONTAINER_NODE_KINDS = %w[binary_expression].freeze
      PARENTHESIZED_WRAPPER_NODE_KINDS = %w[parenthesized_expression].freeze
      ARGUMENT_LIST_NODE_KINDS = %w[arguments argument].freeze
      SELF_CALL_IDENTIFIER_NODE_KINDS = %w[name variable_name].freeze
      SELF_RECEIVER_NAMES = %w[$this this self].freeze
      ACCESSOR_CALL_NODE_KINDS = [].freeze
      FIELD_LIKE_NODE_KINDS = %w[
        member_access_expression nullsafe_member_access_expression member_call_expression
        class_constant_access_expression
      ].freeze
      BLOCK_ARGUMENT_NODE_KINDS = [].freeze

      def function_name(node)
        return php_name_text(named_field(node, "name") || node.named_children.find { |child| child.kind == "name" }) if %w[
          function_definition method_declaration
        ].include?(node.kind)

        super
      end

      def owner_name_from_declaration(document, node)
        return php_name_text(named_field(node, "name") || node.named_children.find { |child| child.kind == "name" }) if node.kind == "class_declaration"

        super
      end

      def visibility(_document, node)
        modifier = node.named_children.find { |child| child.kind == "visibility_modifier" }
        return modifier.text.to_sym if modifier && %w[public private protected].include?(modifier.text)

        :public
      end

      def function_params(node)
        params = named_field(node, "parameters") ||
                 node.named_children.find { |child| child.kind == "formal_parameters" }
        return super unless params

        params.named_children.filter_map { |param| php_parameter_name(param) }.uniq
      end

      def call_target(document, node)
        php_call_target(node) || super
      end

      def state_read_target(node)
        return nil if php_assignment_lhs?(node)

        php_argument_member_target(node) || super
      end

      def state_declaration(node)
        php_property_declaration(node) || super
      end

      def predicate_def(document, function_def)
        predicate = super
        return nil unless predicate

        PredicateDef.new(
          file: predicate.file,
          name: predicate.name,
          owner: predicate.owner,
          body: php_normalize_source(predicate.body),
          line: predicate.line,
          span: predicate.span
        )
      end

      def path_condition_sites(document)
        super.map do |site|
          PathConditionSite.new(
            guards: site.guards.map { |guard| php_normalize_source(guard) },
            action: php_normalize_source(site.action),
            file: site.file,
            function: site.function,
            line: site.line,
            span: site.span
          )
        end
      end

      def local_contract_assignments(document, method)
        super.transform_values { |source| php_normalize_source(source) }
      end

      def redundant_nil_guard_findings(document)
        findings = []
        document.function_defs.each do |function_def|
          php_nil_guard_walk(document, function_def.body, function_def.name, Set.new, findings)
        end
        findings
      end

      private

      def php_call_target(node)
        return php_print_target(node) if node.kind == "print_intrinsic"
        return nil unless %w[function_call_expression member_call_expression scoped_call_expression].include?(node.kind)

        names = node.named_children.select do |child|
          php_name_node?(child) || child.kind == "variable_name" || child.kind == "member_access_expression"
        end
        args = node.named_children.find { |child| child.kind == "arguments" }

        case node.kind
        when "member_call_expression"
          receiver = php_member_receiver(node) || names.first
          message = php_member_name(node) || names[1]
          return nil unless receiver && message

          {
            receiver: php_normalize_receiver(php_identifier_text(receiver)),
            message: php_name_text(message),
            arguments: php_argument_texts(args)
          }
        when "scoped_call_expression"
          receiver = names.first
          message = names[1]
          return nil unless receiver && message

          {
            receiver: php_name_text(receiver),
            message: php_name_text(message),
            arguments: php_argument_texts(args)
          }
        when "function_call_expression"
          name = names.first
          return nil unless name

          {
            receiver: "self",
            message: php_name_text(name),
            arguments: php_argument_texts(args)
          }
        end
      end

      def php_print_target(node)
        {
          receiver: "self",
          message: "print",
          arguments: node.named_children.map { |child| php_print_argument_text(child) }
        }
      end

      def conjunction_predicate(node)
        php_normalize_source(super)
      end

      def branch_predicate(node)
        php_normalize_source(super)
      end

      def php_property_declaration(node)
        return nil unless node.kind == "property_declaration"

        property = node.named_children.find { |child| child.kind == "property_element" }
        name = property&.named_children&.find { |child| child.kind == "variable_name" }
        return nil unless name

        { field: php_identifier_text(name), type: declared_type_text(node, name) }
      end

      def php_parameter_name(param)
        variable = param.named_children.find { |child| child.kind == "variable_name" }
        php_identifier_text(variable) || php_identifier_text(param)
      end

      def generic_identifier?(node)
        super || (ts_node?(node) && %w[name variable_name].include?(node.kind))
      end

      def generic_local_identifier_text(node)
        return php_identifier_text(node) if ts_node?(node) && node.kind == "variable_name"

        super
      end

      def generic_member_name?(node)
        return true if parent_node(node)&.kind == "variable_name"
        return false if node.kind == "variable_name"

        super
      end

      def generic_local_writes(node, **kwargs)
        (super(node, **kwargs) + php_local_write_names(node)).map { |name| php_identifier_text_value(name) }.uniq
      end

      def generic_local_write_node?(node)
        return true if ts_node?(node) && node.kind == "variable_name" && php_assignment_lhs?(node)

        super
      end

      def decision_member_text(node)
        php_normalize_source(super)
      end

      def decision_predicate(node)
        php_normalize_source(super)
      end

      def comparison_target(node)
        target = super
        return nil unless target

        target.merge(source: php_normalize_source(target[:source]))
      end

      def control_context(node)
        return :iterates if node.kind == "foreach_statement"

        super
      end

      def target_from_callee(callee)
        target = super
        return target unless target

        target.merge(receiver: php_normalize_receiver(target[:receiver]))
      end

      def generic_state_read_target(node)
        target = super
        return target unless target

        target.merge(receiver: php_normalize_receiver(target[:receiver]))
      end

      def generic_state_target(lhs)
        target = super
        return target unless target

        target.merge(receiver: php_normalize_receiver(target[:receiver]))
      end

      def member_field_text(field)
        php_name_text(field)
      end

      def simple_identifier_text?(text)
        php_identifier_text_value(text).match?(/\A[A-Za-z_]\w*\z/)
      end

      def php_name_node?(node)
        ts_node?(node) && %w[name qualified_name].include?(node.kind)
      end

      def php_assignment_lhs?(node)
        parent = parent_node(node)
        return false unless parent

        %w[assignment_expression augmented_assignment_expression].include?(parent.kind) &&
          parent.named_children.first == node
      end

      def php_local_write_names(node)
        writes = []
        generic_walk_local(node) do |child|
          next unless ts_node?(child) && child.kind == "variable_name"
          next unless php_assignment_lhs?(child)

          writes << php_identifier_text(child)
        end
        writes.compact
      end

      def php_argument_texts(args)
        Array(args&.named_children).map { |child| php_normalize_source(child.text) }
      end

      def php_print_argument_text(node)
        value = php_unwrap_parenthesized(node)
        php_normalize_source(value&.text || node.text)
      end

      def php_argument_member_target(node)
        return nil unless ts_node?(node) && node.kind == "argument"
        return nil unless node.text.to_s.include?("->") || node.text.to_s.include?("::")
        return nil if node.text.to_s.include?("(")

        parts = php_normalize_source(node.text).split(".")
        return nil unless parts.size >= 2

        {
          receiver: php_normalize_receiver(parts[0...-1].join(".")),
          field: php_identifier_text_value(parts.last)
        }
      end

      def php_member_receiver(node)
        return nil unless ts_node?(node)

        named_field(node, "object") || named_field(node, "receiver") ||
          named_field(node, "expression") || node.named_children.first
      end

      def php_member_name(node)
        return nil unless ts_node?(node)

        named_field(node, "name") || named_field(node, "field") ||
          node.named_children.reverse.find { |child| php_name_node?(child) }
      end

      def php_identifier_text(node)
        text = php_identifier_text_value(node&.text)
        text.empty? ? nil : text
      end

      def php_name_text(node)
        text = php_identifier_text_value(node&.text)
        text.empty? ? nil : text
      end

      def php_identifier_text_value(text)
        text.to_s.sub(/\A\$/, "")
      end

      def php_normalize_receiver(receiver)
        value = php_normalize_source(php_identifier_text_value(receiver))
        value == "this" ? "self" : value
      end

      def php_normalize_source(source)
        source.to_s
              .gsub(/\$([A-Za-z_]\w*)/, '\1')
              .gsub(/->|::/, ".")
      end

      def php_nil_guard_walk(document, node, function, known, findings)
        return unless ts_node?(node)
        return if generic_nested_local_scope?(node) && function_name(node) != function

        if node.kind == "if_statement"
          php_process_nil_guard_if(document, node, function, known, findings)
          return
        end

        php_record_redundant_nil_guard(document, node, function, known, findings)
        node.named_children.each do |child|
          php_nil_guard_walk(document, child, function, known, findings)
        end
      end

      def php_process_nil_guard_if(document, node, function, known, findings)
        condition = named_field(node, "condition") || node.named_children.first
        body = named_field(node, "body") || node.named_children[1]
        branch_known = known.dup
        php_non_nil_facts(condition).each { |local| branch_known.add(local) }
        php_nil_guard_walk(document, body, function, branch_known, findings)
      end

      def php_record_redundant_nil_guard(document, node, function, known, findings)
        subject = php_nil_guard_subject(node)
        return unless subject && known.include?(subject)

        findings << NilGuardFinding.new(
          file: document.file,
          defn: function,
          line: line(node),
          span: span(node),
          local: subject,
          guard: php_normalize_source(node.text),
          proof: "#{subject} is already proven non-nil on this path"
        )
      end

      def php_non_nil_facts(node)
        node = php_unwrap_parenthesized(node)
        return [] unless ts_node?(node)

        subject = php_subject_key(node)
        return [subject] if subject

        call = php_member_call_parts(node)
        return [call[:receiver]] if call && %w[isSome is_some present].include?(call[:message])

        comparison = php_nil_comparison(node)
        return [comparison[:subject]] if comparison && %w[!== !=].include?(comparison[:operator])

        []
      end

      def php_nil_guard_subject(node)
        node = php_unwrap_parenthesized(node)
        return nil unless ts_node?(node)

        call = php_member_call_parts(node)
        return call[:receiver] if call && %w[isNull is_null nil is_none].include?(call[:message])

        comparison = php_nil_comparison(node)
        return comparison[:subject] if comparison && %w[=== ==].include?(comparison[:operator])

        function_call = php_function_call_parts(node)
        return function_call[:arguments].first if function_call && %w[is_null].include?(function_call[:message])

        nil
      end

      def php_nil_comparison(node)
        return nil unless ts_node?(node) && node.kind == "binary_expression"

        operator = direct_operator(node)
        return nil unless %w[=== !== == !=].include?(operator)

        left, right = node.named_children
        if php_null_literal?(right)
          subject = php_subject_key(left)
        elsif php_null_literal?(left)
          subject = php_subject_key(right)
        end
        subject ? { subject: subject, operator: operator } : nil
      end

      def php_member_call_parts(node)
        node = php_unwrap_parenthesized(node)
        return nil unless ts_node?(node) && node.kind == "member_call_expression"

        access = node.named_children.find { |child| child.kind == "member_access_expression" }
        receiver_node = access ? php_member_receiver(access) : node.named_children.find { |child| child.kind == "variable_name" }
        message_node = access ? php_member_name(access) : node.named_children.find { |child| php_name_node?(child) }
        receiver = php_subject_key(receiver_node)
        message = php_name_text(message_node)
        return nil unless receiver && message

        { receiver: receiver, message: message }
      end

      def php_function_call_parts(node)
        node = php_unwrap_parenthesized(node)
        return nil unless ts_node?(node) && node.kind == "function_call_expression"

        name = node.named_children.find { |child| php_name_node?(child) }
        args = node.named_children.find { |child| child.kind == "arguments" }
        {
          message: php_name_text(name),
          arguments: Array(args&.named_children).filter_map { |child| php_subject_key(child) }
        }
      end

      def php_subject_key(node)
        node = php_unwrap_parenthesized(node)
        return nil unless ts_node?(node)

        case node.kind
        when "variable_name", "name"
          php_identifier_text(node)
        when "member_access_expression"
          receiver = php_subject_key(php_member_receiver(node))
          message = php_name_text(php_member_name(node))
          receiver && message ? "#{receiver}.#{message}" : nil
        else
          nil
        end
      end

      def php_unwrap_parenthesized(node)
        current = node
        while ts_node?(current) &&
              %w[parenthesized_expression parenthesized_statements].include?(current.kind) &&
              current.named_children.size == 1
          current = current.named_children.first
        end
        current
      end

      def php_null_literal?(node)
        ts_node?(node) && node.kind == "null"
      end
    end
  end
end
